src/Features/Core/Portable/CodeAnalysisSuggestions/CodeAnalysisSuggestionsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,183 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeAnalysisSuggestions;
+
+[ExportCodeRefactoringProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+internal sealed partial class CodeAnalysisSuggestionsCodeRefactoringProvider
+    : CodeRefactoringProvider
+{
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CodeAnalysisSuggestionsCodeRefactoringProvider()
+    {
+    }
+
+    protected override CodeActionRequestPriority ComputeRequestPriority() => CodeActionRequestPriority.Low;
+
+    public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+    {
+        var (document, span, cancellationToken) = context;
+
+        var configService = document.Project.Solution.Services.GetRequiredService<ICodeAnalysisSuggestionsConfigService>();
+        var ruleConfigData = await configService.TryGetCodeAnalysisSuggestionsConfigDataAsync(document.Project, cancellationToken).ConfigureAwait(false);
+        if (ruleConfigData.IsEmpty)
+            return;
+
+        var codeFixService = document.Project.Solution.Services.ExportProvider.GetExports<ICodeFixService>().FirstOrDefault()?.Value;
+        if (codeFixService == null)
+            return;
+
+        var (actions, totalViolations) = await GetCodeAnalysisSuggestionActionsAsync(ruleConfigData, document, codeFixService, cancellationToken).ConfigureAwait(false);
+        if (actions.Length > 0)
+        {
+            Debug.Assert(totalViolations > 0);
+
+            context.RegisterRefactoring(
+                CodeAction.Create(
+                    string.Format(FeaturesResources.Code_analysis_improvements_0, totalViolations),
+                    actions,
+                    isInlinable: false,
+                    CodeActionPriority.Low),
+                span);
+        }
+    }
+
+    private static async Task<(ImmutableArray<CodeAction> Actions, int TotalViolations)> GetCodeAnalysisSuggestionActionsAsync(
+        ImmutableArray<(string, ImmutableDictionary<string, ImmutableArray<DiagnosticData>>)> configData,
+        Document document,
+        ICodeFixService codeFixService,
+        CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var location = root.GetLocation();
+
+        using var _1 = ArrayBuilder<CodeAction>.GetInstance(out var actionsBuilder);
+        using var _2 = ArrayBuilder<CodeAction>.GetInstance(out var nestedActionsBuilder);
+        using var _3 = ArrayBuilder<CodeAction>.GetInstance(out var nestedNestedActionsBuilder);
+        var totalViolations = 0;
+        foreach (var (category, diagnosticsById) in configData)
```

Comment:
what's the overall costs here?  I'm not really understanding the sizes of these objects, and how much computation will be occurring.  are there scaling concerns here on large projects (like roslyn).

Reply:
Addressed now - we only work off computed/cached diagnostics from background analysis now. A future PR will add a separate nested code action for project wide suggestions such that invoking that will explicitly force compute all project diagnostics and show all project wide suggestions.